src/endpoint.jl

```diff
@@ -21,18 +30,29 @@ const RightEndpoint{T,B} = Endpoint{T, Right, B} where {T,B}
 LeftEndpoint{B}(ep::T) where {T,B} = LeftEndpoint{T,B}(ep)
 RightEndpoint{B}(ep::T) where {T,B} = RightEndpoint{T,B}(ep)
 
-LeftEndpoint(i::AbstractInterval{T,L,R}) where {T,L,R} = LeftEndpoint{T,L}(first(i))
-RightEndpoint(i::AbstractInterval{T,L,R}) where {T,L,R} = RightEndpoint{T,R}(last(i))
+LeftEndpoint(i::AbstractInterval{T,L,R}) where {T,L,R} = LeftEndpoint{T,L}(L !== Unbounded ? first(i) : nothing)
+RightEndpoint(i::AbstractInterval{T,L,R}) where {T,L,R} = RightEndpoint{T,R}(R !== Unbounded ? last(i) : nothing)
 
+endpoint(x::Endpoint) = isbounded(x) ? x.endpoint : nothing
 bound_type(x::Endpoint{T,D,B}) where {T,D,B} = B
+
 isclosed(x::Endpoint) = bound_type(x) === Closed
+isunbounded(x::Endpoint) = bound_type(x) === Unbounded
+isbounded(x::Endpoint) = bound_type(x) !== Unbounded
 
 function Base.hash(x::Endpoint{T,D,B}, h::UInt) where {T,D,B}
-    # Note: we shouldn't need to hash `T` as this is covered by the endpoint field.
+    # Note: we shouldn't need to hash `T` as this is typically covered by the endpoint field
     h = hash(:Endpoint, h)
     h = hash(D, h)
     h = hash(B, h)
-    h = hash(x.endpoint, h)
+
+    # Unbounded endpoints should ignore value.
+    if B !== Unbounded
+        h = hash(x.endpoint, h)
+    else
+        h = hash(T, h)
+    end
```

Comment:
could be one line? `h = B !== Unbounded ? hash(x.endpoint, h) : hash(T, h)`

Reply:
It can be. I thought this was easier to read but I was on the fence